### PR TITLE
CUDA: CUDAHOSTCXX Env

### PR DESCRIPTION
**Pull request by ax3l**
This is a general CMake CUDA language hint to use the CXX compiler has host compiler for NVCC. Seems like a good default since we do not express the CUDA compiler in Spack otherwise yet (e.g. no `self.compiler.cuda` or `self.compiler.cudahostcxx`).

cc @svenevs

- [ ] also add [`CUDACXX`](https://cmake.org/cmake/help/v3.14/envvar/CUDACXX.html)?
- [ ] Maybe also add to mixing class `lib/spack/spack/build_systems/cuda.py`?

### CODE REVIEW

**Comment by svenevs** [comment]
`CUDAHOSTCXX` is one avenue, but for CMake my understanding is it only takes effect if

1. `CC` / `CXX` env vars are **not** set, and
2. Did not configure with `-DCMAKE_{C,CXX}_COMPILER=...`
3. Project uses `CUDA` language support (see below).

I will do some testing later today to verify what the order of preference is.  AKA no particular objection, but it may not be necessary.  I don't think it hurts at all.

----------

TL;DR: I do not believe we are ready for `CUDACXX` yet and should push that off, it will have no meaningful change to any existing software.  The only valid value for currently released CMake is `nvcc`.

CMake  3.18 will add `clang++` support, but "only" with an [unreleased version of LLVM (11)](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4442#note_737136).  Unless spack already has an existing ability to check `libstdc++ > 7`, I'd rather not deal with that and instead only support `clang@11:`.

@ax3l knows what this means, but for the unacquainted, many projects have not yet adopted `CUDA` language support in CMake and are still using the deprecated `find_package(CUDA)` approach.  Meaning setting either of these two env vars will be "meaningless" in terms of affecting the build.  Unfortunately, switching to the new approach (particularly for large projects) is a pretty big cookie, so adoption has been slow.

1. The concretizer will need a way to make it so that the current `+cuda` approach will support using `clang++` compiler.
2. There will be false hope for many packages that this does what is desired.  Though it would be a big maintenance overhead, I think whitelisting packages for non-`nvcc` `CUDACXX` would be prudent.  I help can review CMake packages to determine if they support it if we decide to go this way.

@adamjstewart and I spoke about ways to update the constraint logic for cuda (it's functional, but kind of convoluted).  He said that after the new concretizer is in we should be able to make that more flexible.  I think it would make sense to add in LLVM capabilities with that, depending on how soon we need it (I don't think LLVM 11 comes out for a while?).  Additionally, CUDA 11 will (_finally_) be adding an allow unsupported flag to nvcc (meaning `^cuda@11` won't need a hard `conflicts` anymore).

> Maybe also add to mixing class `lib/spack/spack/build_systems/cuda.py`

I don't know what that means (I'm not too good with the internals of spack), but it sounds like it's progress toward (1)?

**Comment by ax3l** [comment]
Just a minor addition: `CUDACXX` takes effect either way, independent of `CC` and `CXX`. It's the CUDA language equivalent of `CMAKE_<LANG>_COMPILER` and as always, arguments take precedence over environment vars. This is where `nvcc` (or later `clang -x cuda`) go.

`CUDACXX`/`CMAKE_CUDA_COMPILER` are relevant if one has several CUDA versions installed, e.g. one in system paths and one via Spack. Probably Spack prepends paths, do we make other language compiler envs/hints explicit, too? I think we do with the `SPACK_FC`/`SPACK_CC`/`SPACK_CXX` wrappers.

`CUDAHOSTCXX`/`CMAKE_CUDA_HOST_COMPILER ` is where a host compiler for the CUDA language goes. It is intentionally not default initialized with `CXX`/`CMAKE_CXX_COMPILER` because the CMake developers (paraphrasing @robertmaynard here) wanted to keep individual languages side-effect free. In fact, _C++_ does not even need to be enabled as a language to use the CMake _CUDA C++_ language.

I think as of concern for this PR, I am good to just add `CUDAHOSTCXX` for now. This is where I encounter real-life problems with compiler selection as of today.

**Comment by robertmaynard** [comment]
> Just a minor addition: `CUDACXX` takes effect either way, independent of `CC` and `CXX`. It's the CUDA language equivalent of `CMAKE_<LANG>_COMPILER` and as always, arguments take precedence over environment vars. This is where `nvcc` (or later `clang -x cuda`) go.
>
> `CUDAHOSTCXX`/`CMAKE_CUDA_HOST_COMPILER ` is where a host compiler for the CUDA language goes. It is intentionally not default initialized with `CXX`/`CMAKE_CXX_COMPILER` because the CMake developers (paraphrasing @robertmaynard here) wanted to keep individual languages side-effect free. In fact, _C++_ does not even need to be enabled as a language to use the CMake _CUDA C++_ language.
> 

Correct

**Review by svenevs — APPROVED** [review]
Awesome, thanks for following up @ax3l and for confirming @robertmaynard . I couldn't figure out the relation with CUDAHOSTCXX with certainty, sorry for not responding at least saying that.

Given the discussion above I think this looks good as is, we can tackle the other stuff at a later date.

**Review by adamjstewart — APPROVED** [review]
